### ISSUE TITLE
6.0.5: pytest fails in 7 units

### ISSUE THREAD

**Issue by kloczek** [aio-libs/multidict#959]
I'm packaging your module as an rpm package so I'm using the typical PEP517 based build, install and test cycle used on building packages from non-root account.
- `python3 -sBm build -w --no-isolation`
- because I'm calling `build` with `--no-isolation` I'm using during all processes only locally installed modules
- install .whl file in </install/prefix> using `installer` module
- run pytest with $PYTHONPATH pointing to sitearch and sitelib inside </install/prefix>
- build is performed in env which is *`cut off from access to the public network`* (pytest is executed with `-m "not network"`)

<details>
<summary>Here is pytest output:</summary>

```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-multidict-6.0.5~no_loopy_deps-2.fc36.x86_64/usr/lib64/python3.9/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-multidict-6.0.5~no_loopy_deps-2.fc36.x86_64/usr/lib/python3.9/site-packages
+ /usr/bin/pytest -ra -m 'not network'
============================= test session starts ==============================
platform linux -- Python 3.9.18, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/tkloczko/rpmbuild/BUILD/multidict-6.0.5
configfile: pytest.ini
testpaths: tests/
collected 1002 items

tests/test_abc.py .........                                              [  0%]
tests/test_copy.py ....                                                  [  1%]
tests/test_multidict.py ...                                              [  1%]
tests/test_pickle.py ......                                              [  2%]
tests/test_abc.py ..                                                     [  2%]
tests/test_copy.py ....                                                  [  2%]
tests/test_multidict.py ...                                              [  3%]
tests/test_pickle.py .........................................           [  7%]
tests/test_update.py ..........                                          [  8%]
tests/test_version.py ...................                                [ 10%]
tests/test_pickle.py ..                                                  [ 10%]
tests/test_abc.py ..                                                     [ 10%]
tests/test_copy.py ....                                                  [ 10%]
tests/test_multidict.py ...                                              [ 11%]
tests/test_pickle.py .                                                   [ 11%]
tests/test_update.py ..........                                          [ 12%]
tests/test_version.py ...................                                [ 14%]
tests/test_multidict.py .                                                [ 14%]
tests/test_abc.py ..                                                     [ 14%]
tests/test_copy.py ....                                                  [ 14%]
tests/test_multidict.py ...                                              [ 15%]
tests/test_pickle.py .                                                   [ 15%]
tests/test_update.py ..........                                          [ 16%]
tests/test_version.py ...................                                [ 18%]
tests/test_guard.py ...                                                  [ 18%]
tests/test_istr.py ........                                              [ 19%]
tests/test_multidict.py ................................................ [ 24%]
........................................................................ [ 31%]
........................................................................ [ 38%]
........................................................................ [ 45%]
......................................                                   [ 49%]
tests/test_mutable_multidict.py ........................................ [ 53%]
...........                                                              [ 54%]
tests/test_types.py ............s.                                       [ 55%]
tests/test_update.py ...                                                 [ 56%]
tests/test_version.py .                                                  [ 56%]
tests/test_pickle.py .                                                   [ 56%]
tests/test_update.py ..........                                          [ 57%]
tests/test_version.py ...................                                [ 59%]
tests/test_multidict.py .                                                [ 59%]
tests/test_guard.py ...                                                  [ 59%]
tests/test_istr.py ........                                              [ 60%]
tests/test_multidict.py ................................................ [ 65%]
........................................................................ [ 72%]
........................................................................ [ 79%]
........................................................................ [ 86%]
......................................                                   [ 90%]
tests/test_mutable_multidict.py ........................................ [ 94%]
...........                                                              [ 95%]
tests/test_types.py ............s.                                       [ 97%]
tests/test_update.py ...                                                 [ 97%]
tests/test_version.py .                                                  [ 97%]
tests/test_abc.py .                                                      [ 97%]
tests/test_circular_imports.py FFFFFFF                                   [ 98%]
tests/test_mypy.py .................                                     [100%]

=================================== FAILURES ===================================
_________________________ test_no_warnings[multidict] __________________________

import_path = 'multidict'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict')
import_path = 'multidict'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
_______________________ test_no_warnings[multidict._abc] _______________________

import_path = 'multidict._abc'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc')
import_path = 'multidict._abc'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._abc'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
_____________________ test_no_warnings[multidict._compat] ______________________

import_path = 'multidict._compat'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat')
import_path = 'multidict._compat'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._compat'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
____________________ test_no_warnings[multidict._multidict] ____________________

import_path = 'multidict._multidict'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict')
import_path = 'multidict._multidict'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
_________________ test_no_warnings[multidict._multidict_base] __________________

import_path = 'multidict._multidict_base'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base')
import_path = 'multidict._multidict_base'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_base'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
__________________ test_no_warnings[multidict._multidict_py] ___________________

import_path = 'multidict._multidict_py'

    def test_no_warnings(import_path: str) -> None:
        """Verify that importing modules and packages doesn't explode.

        This is seeking for any import errors including ones caused
        by circular imports.
        """
        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", f"import {import_path!s}",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py')
import_path = 'multidict._multidict_py'

tests/test_circular_imports.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py'),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py')

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py')' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py')
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', 'import multidict._multidict_py'),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
____________________ test_c_extension_preferred_by_default _____________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f942949a8b0>

    @pytest.mark.c_extension
    def test_c_extension_preferred_by_default(monkeypatch: pytest.MonkeyPatch) -> None:
        """Verify that the C-extension is exposed by default."""
        monkeypatch.delenv("MULTIDICT_NO_EXTENSIONS", raising=False)

        imp_cmd = (
            # fmt: off
            sys.executable,
            "-I",
            "-W", "error",
            "-c", "import multidict; raise SystemExit(int("
            "multidict.istr.__module__ != 'multidict._multidict' "
            "or multidict.USE_EXTENSIONS is not True))",
            # fmt: on
        )

>       subprocess.check_call(imp_cmd)

imp_cmd    = ('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))")
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f942949a8b0>

tests/test_circular_imports.py:112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))"),)
kwargs = {}, retcode = 1
cmd = ('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))")

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.

        The arguments are the same as for the call function.  Example:

        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))")' returned non-zero exit status 1.

cmd        = ('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))")
kwargs     = {}
popenargs  = (('/usr/bin/python3', '-I', '-W', 'error', '-c', "import multidict; raise SystemExit(int(multidict.istr.__module__ != 'multidict._multidict' or multidict.USE_EXTENSIONS is not True))"),)
retcode    = 1

/usr/lib64/python3.9/subprocess.py:373: CalledProcessError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'multidict'
- generated xml file: /home/tkloczko/rpmbuild/BUILD/multidict-6.0.5/junit-test-results.xml -
============================= slowest 10 durations =============================
0.10s call     tests/test_mutable_multidict.py::TestMutableMultiDict::test_large_multidict_resizing[pure-python-module]
0.06s teardown tests/test_mypy.py::test_popall
0.04s call     tests/test_istr.py::test_leak[c-extension-module]
0.03s call     tests/test_istr.py::test_leak[pure-python-module]
0.02s call     tests/test_mutable_multidict.py::TestCIMutableMultiDict::test_sizeof[pure-python-module]
0.02s call     tests/test_circular_imports.py::test_no_warnings[multidict._abc]
0.02s call     tests/test_circular_imports.py::test_no_warnings[multidict._compat]
0.02s call     tests/test_circular_imports.py::test_no_warnings[multidict._multidict_base]
0.02s call     tests/test_circular_imports.py::test_c_extension_preferred_by_default
0.02s call     tests/test_circular_imports.py::test_no_warnings[multidict._multidict]
=========================== short test summary info ============================
SKIPPED [2] tests/test_types.py:88: Python 3.9 uses GenericAlias which is different
FAILED tests/test_circular_imports.py::test_no_warnings[multidict] - subproce...
FAILED tests/test_circular_imports.py::test_no_warnings[multidict._abc] - sub...
FAILED tests/test_circular_imports.py::test_no_warnings[multidict._compat] - ...
FAILED tests/test_circular_imports.py::test_no_warnings[multidict._multidict]
FAILED tests/test_circular_imports.py::test_no_warnings[multidict._multidict_base]
FAILED tests/test_circular_imports.py::test_no_warnings[multidict._multidict_py]
FAILED tests/test_circular_imports.py::test_c_extension_preferred_by_default
=================== 7 failed, 993 passed, 2 skipped in 2.74s ===================
```
</details>

<details>
<summary>List of installed modules in build env:</summary>

```console
Package            Version
------------------ -----------
build              1.1.1
Cython             3.0.9
distro             1.9.0
dnf                4.19.0
exceptiongroup     1.1.3
gpg                1.23.2
importlib_metadata 7.0.1
iniconfig          2.0.0
installer          0.7.0
libdnf             0.73.0
packaging          24.0
pluggy             1.4.0
pyproject_hooks    1.0.0
pytest             8.1.1
python-dateutil    2.9.0.post0
setuptools         69.1.1
tokenize_rt        5.2.0
tomli              2.0.1
typing_extensions  4.10.0
wheel              0.43.0
zipp               3.17.0
```
</details>

Please let me know if you need more details or want me to perform some diagnostics.

**Comment by kloczek** [comment]
Just checked and this module has some DSOs and looks like python code uses relative imports which always causes that kind of issues.
Will try to prepare patch which removes use relative imports and test is it help.

**Comment by webknjaz** [comment]
This is not a problem. The tests use `-I` which drops your custom `$PYTHONPATH`. If you install properly, this won't be happening.

**Comment by kloczek** [comment]
Another though .. why those units are trying to import DSOs modules from python command line?
They've already have been tested by other units that it is possible to import them 🤔

**Comment by webknjaz** [comment]
Because circular imports must be tested in isolation and not rely on import caching from other tests.

**Comment by kloczek** [comment]
> This is not a problem. The tests use `-I` which drops your custom `$PYTHONPATH`. If you install properly, this won't be happening.

Procedure which I've described opening ticket is typically used on packaging python modules as rpm packages. 😞 
In other words those units are breaking "testing as installed" methodology 🤔

**Comment by webknjaz** [comment]
Also, I have no idea what you mean by DSO.

**Comment by webknjaz** [comment]
> In other words those units are breaking \ñ"testing as installed" methodology 🤔

"testing as installed" is exactly what said tests are doing. Your tests are wrong and test "as injected by an arbitrary env var", which is different. We don't want accidental imports from the source directory.

**Comment by kloczek** [comment]
> Because circular imports must be tested in isolation and not rely on import caching from other tests.

IIRC cruelty pytest always test circular imports 🤔 
This is why none of the other modules test suites are doing what those units here are trying to do.

**Comment by kloczek** [comment]
> Also, I have no idea what you mean by DSO.

DSO stands for Dynamic Shared Object.

**Comment by webknjaz** [comment]
> This is why none of the other modules test suites are doing what those units here are trying to do.

That's entirety untrue. Many of my projects do this. Moreover, I borrowed the idea from pytest's own test suite.

**Comment by webknjaz** [comment]
> python code uses relative imports which always causes that kind of issues.

This is also totally false, by the way.

**Comment by kloczek** [comment]
> This is also totally false, by the way.

Can I read somewhere about that and/or can you explain that? 🤔 
AFAIK uses relative imports causes:
- additional syscall to obtain current directory
- breaks testing on typical build procedures used on packaging causing to use additional venvs to test module code/DSOs.

**Comment by kloczek** [comment]
> That's entirety untrue. Many of my projects do this. Moreover, I borrowed the idea from pytest's own test suite.

I have packaged ~1.24k modules as rpm packages. Many of them comes with own DSOs as well.
If that methodology is so widely spread I would be able to observe something which you have here in those test unis.
So far this case is first one and I must admit that I'm testing almost those modules using pytest (+99%)
```console
[tkloczko@pers-jacek SPECS]$ grep -l ^%pytest python-*.spec|wc -l; ls -1 python-*.spec|wc -l
1220
1238
```
I have as well multiple of your modules
```console
[tkloczko@pers-jacek SPECS]$ grep https://github.com/aio-libs/ *spec -l
python-aiobotocore.spec
python-aiohttp-cors.spec
python-aiohttp.spec
python-aiopg.spec
python-aioredis.spec
python-aiosignal.spec
python-async-lru.spec
python-async-timeout.spec
python-frozenlist.spec
python-multidict.spec
python-pytest-aiohttp.spec
python-sphinxcontrib-asyncio.spec
python-yarl.spec
```
so it is a bit odd that only here I was able to spot failing units on checking circular dependencies methodology which is used here.
I have already multiple tickets opened in your repos 
```console
[tkloczko@pers-jacek SPECS]$ grep https://github.com/aio-libs/ *spec |grep BUG
python-aiobotocore.spec:# BUG: pytest fais https://github.com/aio-libs/aiobotocore/issues/965
python-aiohttp-cors.spec:# BUG: test suite fails https://github.com/aio-libs/aiohttp-cors/issues/383
python-aiohttp.spec:# BUG: pytest fais https://github.com/aio-libs/aiohttp/issues/7255
python-aiohttp.spec:# BUG: sphinx fails https://github.com/aio-libs/aiohttp/issues/7230
python-aiopg.spec:# BUG: sphinx fails https://github.com/aio-libs/aiopg/issues/898
python-aioredis.spec:# BUG: pytest is https://github.com/aio-libs/aioredis-py/issues/1406
python-async-timeout.spec:# BUG: test suite warnings https://github.com/aio-libs/async-timeout/issues/203
python-frozenlist.spec:# BUG: sphinx fails https://github.com/aio-libs/frozenlist/issues/570
python-sphinxcontrib-asyncio.spec:# BUG: missing test suite? https://github.com/aio-libs/sphinxcontrib-asyncio/issues/9
python-yarl.spec:# BUG: build fails https://github.com/aio-libs/yarl/issues/969
python-yarl.spec:# BUG: sphinx is failing https://github.com/aio-libs/yarl/issues/970
```
and I do not remember anything similar (.. maybe my memory tricks me).

**Comment by kloczek** [comment]
> > python code uses relative imports which always causes that kind of issues.
> 
> This is also totally false, by the way.

I many in this case I've been trying to tell that generally use relative imports causes problems with importing modules on use "test as installed" methodology used on packaging.

**Comment by webknjaz** [comment]
> > > python code uses relative imports which always causes that kind of issues.
> > 
> > This is also totally false, by the way.
> 
> I many in this case I've been trying to tell that generally use relative imports causes problems with importing modules on use "test as installed" methodology used on packaging.
> 

I don't think you understand how imports work but I also don't have time to educate you so you'll just have to take my word for it 🤷‍♂️

**Comment by kloczek** [comment]
> This is not related to C-extensions. I don't know why you keep bringing it up. It's also not related to relative imports.

If DSO module is imported from some pyton file by `import .foo` it needs to be located at the same directory where is that file.
pep517 AFAIK do not support building module in-place to put linked DSO where is .py file.
On packaging as far as someone don't want to use venv it meas that it cannot be used methodology which I've described on top of this ticket.

BTW:
>  We don't want accidental imports from the source directory.

As long as is used pytest wrapper script instead `python -m pytest` on pytest execution is not used module code from source tree (`python -m foo` adds to `sys.paths` current directory). Only bits which are used are test suite files (test*py files found by pytest on scanning units).
Please have look on your CI
https://github.com/aio-libs/multidict/blob/8a591caf64abcd28ab93f939139f286aefdecf92/.github/workflows/ci-cd.yml#L298
in other words you are exposing yourself on that kind of situation.
It was many times repeated on pytest forum to NOT use `python -m pytest`.
Another thing about that line. Instead passing `tests` as param better is to have defined `testpaths` in pytest settings (if it is already defined that param may be redundant here).
